pennylane/devices/qubit/preprocess.py

```diff
@@ -40,30 +44,40 @@
     "Evolution",
 }
 
+### UTILITY FUNCTIONS FOR EXPANDING UNSUPPORTED OPERATIONS ###
 
-def _stopping_condition(op):
-    """Specify whether or not an Operator object is supported by the device"""
-    return getattr(op, "has_matrix", False)
 
+def _accepted_operator(op: qml.operation.Operator) -> bool:
+    """Specify whether or not an Operator object is supported by the device."""
+    if op.name == "QFT" and len(op.wires) >= 6:
+        return False
+    if op.name == "GroverOperator" and len(op.wires) >= 13:
+        return False
+    return op.has_matrix
 
-def _supports_observable(observable):
-    """Checks if an observable is supported by this device.
 
-    Args:
-        observable (type or str): observable to be checked
+def _operator_decomposition_gen(
+    op: qml.operation.Operator,
+) -> Generator[qml.operation.Operator, None, None]:
+    """A generator that yields the next operation that is accepted by DefaultQubit2."""
+    if _accepted_operator(op):
+        yield op
+    else:
+        try:
+            decomp = op.decomposition()
+        except qml.operation.DecompositionUndefinedError as e:
+            raise DeviceError(
+                f"Operator {op} not supported on DefaultQubit2. Must provide either a matrix or a decomposition."
+            ) from e
 
-    Returns:
-        bool: ``True`` iff supplied observable is supported
-    """
-    if isinstance(observable, type) and issubclass(observable, Observable):
-        observable = observable.__name__
-    if isinstance(observable, str):
-        return observable in _observables
+        for sub_op in decomp:
+            yield from _operator_decomposition_gen(sub_op)
 
-    return False
 
+#######################
 
-def expand_fn(circuit, max_expansion=10):
+
+def expand_fn(circuit: qml.tape.QuantumScript) -> qml.tape.QuantumScript:
```

Comment:
Is this `expand_fn` specific to the new device/execution pipeline?

Reply:
Implementation detail specific to `DefaultQubit2`.  Maybe we could rename it to something like `single_circuit_transformation` or something like that.